File: 151051/Makovetskiy/task310+320/dc-rest/src/main/java/com/rmakovetskij/dc_rest/service/IssueService.java
package com.rmakovetskij.dc_rest.service;

import com.rmakovetskij.dc_rest.mapper.IssueMapper;
import com.rmakovetskij.dc_rest.model.Issue;
import com.rmakovetskij.dc_rest.model.Editor;
import com.rmakovetskij.dc_rest.model.dto.requests.IssueRequestTo;
import com.rmakovetskij.dc_rest.model.dto.responses.IssueResponseTo;
import com.rmakovetskij.dc_rest.repository.interfaces.IssueRepository;
import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
public class IssueService {

    private final IssueRepository issueRepository;
    private final IssueMapper issueMapper;
    public IssueResponseTo createIssue(IssueRequestTo issueRequestDto) {
        Issue issue = issueMapper.toEntity(issueRequestDto);

        // Устанавливаем пользователя по ID, переданному в DTO
        Editor editor = new Editor();  // Создаем экземпляр пользователя
        editor.setId(issueRequestDto.getEditorId());  // Устанавливаем ID из DTO
        issue.setEditor(editor);  // Устанавливаем пользователя в историю

        issue = issueRepository.save(issue);

        return issueMapper.toResponse(issue);
    }

    // Получить историю по id
    public IssueResponseTo getIssueById(Long id) {
        Optional<Issue> issueOptional = issueRepository.findById(id);
        return issueOptional.map(issueMapper::toResponse)
                .orElseThrow(() -> new RuntimeException("Issue not found"));
    }

    // Получить все истории
    public List<IssueResponseTo> getAllIssues() {
        return issueRepository.findAll().stream()
                .map(issueMapper::toResponse)
                .toList();
    }

    public IssueResponseTo updateIssue(Long id, IssueRequestTo issueRequestDto) {
        Issue existingIssue = issueRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Issue not found"));

        // Проверка длины заголовка
        if (issueRequestDto.getTitle().length() < 2) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Title must be at least 2 characters long");
        }

        existingIssue.setTitle(issueRequestDto.getTitle());
        existingIssue.setContent(issueRequestDto.getContent());

        issueRepository.save(existingIssue);
        return issueMapper.toResponse(existingIssue);
    }

    // Удалить историю по id
    public void deleteIssue(Long id) {
        if (!issueRepository.existsById(id)) {
            throw new RuntimeException("Issue not found");
        }
        issueRepository.deleteById(id);
    }
}

File: 251003/Pankratiev/rest/publisher/src/main/kotlin/bsuir/dc/publisher/service/IssueService.kt
package bsuir.dc.publisher.service

import bsuir.dc.publisher.dto.from.IssueRequestTo
import bsuir.dc.publisher.dto.to.IssueResponseTo
import bsuir.dc.publisher.entity.Issue
import bsuir.dc.publisher.entity.Label
import bsuir.dc.publisher.entity.Writer
import bsuir.dc.publisher.mapper.toEntity
import bsuir.dc.publisher.mapper.toResponse
import bsuir.dc.publisher.repository.IssueRepository
import bsuir.dc.publisher.repository.WriterRepository
import jakarta.persistence.criteria.Predicate
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.CachePut
import org.springframework.cache.annotation.Cacheable
import org.springframework.cache.annotation.Caching
import org.springframework.data.jpa.domain.Specification
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class IssueService(
    private val issueRepository: IssueRepository,
    private val writerRepository: WriterRepository,
) {
    @Caching(
        evict = [
            CacheEvict(value = ["issues"], key = "'all_issues'"),
        ]
    )
    fun createIssue(issueRequestTo: IssueRequestTo): IssueResponseTo {
        val writer = writerRepository.findById(issueRequestTo.writerId).orElseThrow { NoSuchElementException() }
        val labels = issueRequestTo.labels.map { Label(name = it) }.toMutableSet()
        val issue = issueRequestTo.toEntity(writer, labels)
        val savedIssue = issueRepository.save(issue)
        return savedIssue.toResponse()
    }

    @Cacheable(value = ["issues"], key = "#id")
    fun getIssueById(id: Long): IssueResponseTo {
        val issue = issueRepository.findById(id).orElseThrow { NoSuchElementException() }
        return issue.toResponse()
    }

    @Cacheable(value = ["issues"], key = "'all_issues'")
    fun getAllIssues(): List<IssueResponseTo> =
        issueRepository.findAll().map { it.toResponse() }

    @Caching(
        put = [CachePut(value = ["issues"], key = "#id")],
        evict = [CacheEvict(value = ["issues"], key = "'all_issues'")]
    )
    fun updateIssue(id: Long, issueRequestTo: IssueRequestTo): IssueResponseTo {
        val writer = writerRepository.findById(issueRequestTo.writerId).orElseThrow { NoSuchElementException() }
        val labels = issueRequestTo.labels.map { Label(name = it) }.toMutableSet()
        val updatedIssue = issueRequestTo.toEntity(writer, labels).apply {
            this.id = id
            this.modified = LocalDateTime.now()
        }
        return issueRepository.save(updatedIssue).toResponse()
    }

    @Caching(
        evict = [
            CacheEvict(value = ["issues"], key = "#id"),
            CacheEvict(value = ["issues"], key = "'all_issues'")
        ]
    )
    fun deleteIssue(id: Long) {
        issueRepository.findById(id).orElseThrow { NoSuchElementException() }
        issueRepository.deleteById(id)
    }

    fun getIssuesByFilters(
        labelNames: List<String>?,
        labelIds: List<Long>?,
        writerLogin: String?,
        title: String?,
        content: String?
    ): List<IssueResponseTo> {
        val issueIdsByNames = getIssueIdsByLabelNames(labelNames)
        val issueIdsByLabelIds = getIssueIdsByLabelIds(labelIds)
        val writerId = getWriterIdByLogin(writerLogin)

        return filterIssues(issueIdsByNames, issueIdsByLabelIds, writerId, title, content)
    }

    private fun getIssueIdsByLabelNames(labelNames: List<String>?): Set<Long>? {
        if (labelNames.isNullOrEmpty()) return null

        val lowerCaseNames = labelNames.map { it.lowercase() }
        return issueRepository
            .findByAllLabelNames(lowerCaseNames, lowerCaseNames.count().toLong())
            .map { it.id }.toSet()
    }

    private fun getIssueIdsByLabelIds(labelIds: List<Long>?): Set<Long>? {
        if (labelIds.isNullOrEmpty()) return null

        return issueRepository
            .findByAllLabelIds(labelIds, labelIds.count().toLong())
            .map { it.id }.toSet()
    }

    private fun getWriterIdByLogin(writerLogin: String?): Long? {
        if (writerLogin.isNullOrBlank()) return null

        val writer = writerRepository.findByLoginIgnoreCase(writerLogin)
        return writer?.id ?: -1
    }

    private fun filterIssues(
        issueIdsByNames: Set<Long>?,
        issueIdsByLabelIds: Set<Long>?,
        writerId: Long?,
        title: String?,
        content: String?
    ): List<IssueResponseTo> {
        val specification = IssueSpecifications.filterByParams(issueIdsByNames, issueIdsByLabelIds, writerId, title, content)
        return issueRepository.findAll(specification).map { it.toResponse() }
    }
}

object IssueSpecifications {
    fun filterByParams(
        issueIdsByNames: Set<Long>?,
        issueIdsByLabelIds: Set<Long>?,
        writerId: Long?,
        title: String?,
        content: String?
    ): Specification<Issue> {
        return Specification { root, query, criteriaBuilder ->
            val predicates = mutableListOf<Predicate>()

            issueIdsByNames?.let {
                predicates.add(root.get<Long>("id").`in`(it))
            }

            issueIdsByLabelIds?.let {
                predicates.add(root.get<Long>("id").`in`(it))
            }

            writerId?.let {
                predicates.add(criteriaBuilder.equal(root.get<Writer>("writer").get<Long>("id"), it))
            }

            title?.let {
                predicates.add(criteriaBuilder.like(criteriaBuilder.lower(root.get("title")), "%${it.lowercase()}%"))
            }

            content?.let {
                predicates.add(criteriaBuilder.like(criteriaBuilder.lower(root.get("content")), "%${it.lowercase()}%"))
            }

            criteriaBuilder.and(*predicates.toTypedArray())
        }
    }
}
